sharktank/tests/evaluate/perplexity.py

```diff
@@ -0,0 +1,205 @@
+    """
+
+    def __init__(
+        self,
+        prompts: list,
+    ):
+        self.prompts = prompts
+        self.add_start_token = False
+        self.batch_size = 16
+        self.bs = len(prompts)
+
+    def load_model(self, dataset, tokenizer):
+
+        theta = dataset.root_theta
+
+        config = LlamaModelConfig(
+            hp=configs.LlamaHParams.from_gguf_props(dataset.properties),
+            block_seq_stride=16,
+            kv_cache_type=args.kv_cache_type,
+            device=device,
+            activation_dtype=torch.float16,
+            attention_dtype=torch.float16,
+        )
+
+        if config.hp.expert_count:
+            # TODO Add grok after merging grok changes
+            # if config.hp.model_arch == "grok":
+            # pass
+            # model = PagedGrokModelV1(theta, config)
+            # else:
+            model = PagedMixtralModelV1(theta, config)
+        else:
+            model = PagedLlamaModelV1(theta, config)
+
+        self.generator = TorchGenerator(model, tokenizer)
+
+    def get_logits(self):
+
+        # print(f":: Prompting:")
+        # for prompt in prompts:
+        #     print(f"    {prompt.encode()}")
+
+        token_ids, seq_lens = self.generator.tokenizer.encode(
+            self.prompts,
+            pad_to_multiple_of=self.generator.model.cache.pad_sequence_stride,
+            add_start_token=self.add_start_token,
+        )
+
+        # print(f":: Prompt tokens: {token_ids}")
+        max_prompt_length = max(seq_lens)
+
+        self.token_ids = torch.tensor(token_ids, device=self.generator.model.device)
+        self.attention_mask = (self.token_ids != 0).int()
+
+        is_first_token = True
+        for i in tqdm(
+            range(0, max_prompt_length),
+            desc="eval-perplexity: Load models & Fetching logits",
+        ):
+
+            token_batch = self.token_ids[:, i : i + 1]
+            seq_lens_batch = torch.tensor([i] * self.bs)
+            # print('loop', i, token_batch, seq_lens_batch)
+
+            if is_first_token:
+                token_batch, seq_lens_batch = self.generator.tokenizer.pad_tokens(
+                    token_ids=token_batch.tolist(),
+                    pad_to_multiple_of=self.generator.model.cache.pad_sequence_stride,
+                )
+                token_batch = torch.tensor(
+                    token_batch, device=self.generator.model.device
+                )
+                seq_lens_batch = torch.tensor(
+                    seq_lens_batch, device=self.generator.model.device
+                )
+
+            self.batch = self.generator.begin_eval_batch(
+                token_batch=token_batch,
+                seq_lens_batch=seq_lens_batch,
+                bs=self.bs,
+            )
+
+            if is_first_token:
+                is_first_token = False
+                self.cache_state = self.batch.prefill()
+                # print(self.batch.detokenize())
```

Comment:
Just another example in these prints. If we have a config setting we can optionally print these which would be handy.

Reply:
Yes, the prints are here for debugging, I will cleanup once I validate the run. Having a config setting to print is actually a great idea.